java/src/main/native/src/NvcompJni.cpp

```diff
@@ -0,0 +1,555 @@
+                                                             jlong compressed_sizes_out_ptr,
+                                                             jlongArray in_ptrs,
+                                                             jlongArray in_sizes,
+                                                             jlong chunk_size,
+                                                             jlong temp_ptr,
+                                                             jlong temp_size,
+                                                             jlongArray out_ptrs,
+                                                             jlongArray out_sizes,
+                                                             jlong jstream) {
+  try {
+    cudf::jni::auto_set_device(env);
+    cudf::jni::native_jpointerArray<void const> input_ptrs(env, in_ptrs);
+    cudf::jni::native_jlongArray input_jsizes(env, in_sizes);
+    if (input_ptrs.size() != input_jsizes.size()) {
+      cudf::jni::throw_java_exception(env, NVCOMP_ERROR_CLASS, "input array size mismatch");
+    }
+    std::vector<size_t> input_sizes;
+    std::transform(input_jsizes.data(), input_jsizes.data() + input_jsizes.size(),
+                   std::back_inserter(input_sizes),
+                   [](jlong x) -> size_t { return static_cast<size_t>(x); });
+
+    cudf::jni::native_jpointerArray<void> output_ptrs(env, out_ptrs);
+    cudf::jni::native_jlongArray output_jsizes(env, out_sizes);
+    if (output_ptrs.size() != output_jsizes.size()) {
+      cudf::jni::throw_java_exception(env, NVCOMP_ERROR_CLASS, "output array size mismatch");
+    }
+    if (input_ptrs.size() != output_ptrs.size()) {
+      cudf::jni::throw_java_exception(env, NVCOMP_ERROR_CLASS, "input/output array size mismatch");
+    }
+
+    auto output_sizes = reinterpret_cast<size_t*>(compressed_sizes_out_ptr);
+    std::transform(output_jsizes.data(), output_jsizes.data() + output_jsizes.size(),
+                   output_sizes,
+                   [](jlong x) -> size_t { return static_cast<size_t>(x); });
+
+    nvcompLZ4FormatOpts opts{};
+    opts.chunk_size = chunk_size;
+    auto stream = reinterpret_cast<cudaStream_t>(jstream);
+    auto status = nvcompBatchedLZ4CompressAsync(input_ptrs.data(), input_sizes.data(),
+                                                input_ptrs.size(), &opts,
+                                                reinterpret_cast<void*>(temp_ptr),
+                                                static_cast<size_t>(temp_size),
+                                                output_ptrs.data(),
+                                                output_sizes,  // input/output parameter
+                                                stream);
+    check_nvcomp_status(env, status);
+  } CATCH_STD(env, );
+}
+
+JNIEXPORT jlong JNICALL
+Java_ai_rapids_cudf_nvcomp_NvcompJni_cascadedCompressGetTempSize(JNIEnv *env, jclass,
```

Comment:
It would be nice to have a comment describing what RLE, delta, and bp are. I assume the first one is run-length-encoding, though not sure why we need a number of them. I think that's once per buffer handed to nvcomp.

Reply:
This is covered in the Java docs to the NvcompJni native methods which calls these.